### PR TITLE
WT-2402 Pad structures to avoid cache line sharing.

### PR DESCRIPTION
**Pull request by michaelcahill**
We previously used compiler alignment to avoid cache line sharing, which
is problematic because we don't allocate memory with the necessary
alignment, so some behavior is undefined.  Instead, this change merely
pads structs to ensure they won't share a cache line when allocated in
an array.

### CODE REVIEW

**Comment by michaelcahill** [comment]
NOT READY FOR MERGE -- this is merely a thought bubble for now.

**Comment by michaelcahill** [comment]
Motivation: http://build.wiredtiger.com:8080/job/wiredtiger-sanitize-undefined/214/

**Comment by daveh86** [comment]
I'm looking into this a little more now. It looks like the anonymous unions used in the patch have been made a part of the C11 standard. One of the worries here would have been about the portability of this change.

**Comment by daveh86** [comment]
@keithbostic, can you have a look at this? The change here is very much still WIP and I would like to hear your thoughts on what direction to take.

**Comment by keithbostic** [comment]
@daveh86, this seems fine to me.

Thinking about the change, it seems to me we can reasonably not worry about older compilers and focus on creating good results with current compiler releases.

Some comments:

I had to add `build_posix/aclocal/ax_check_compile_flag.m4` to the tree to try this branch, I don't think it's part of a standard autoconf/automake install.

I'd be inclined to make the `std=c11` test non-fatal, updating AM_CFLAGS` is better, and "+=" doesn't work with some shells:
```
diff --git a/build_posix/configure.ac.in b/build_posix/configure.ac.in
index 5b1327f..90bd8cc 100644
--- a/build_posix/configure.ac.in
+++ b/build_posix/configure.ac.in
@@ -22,6 +22,8 @@ AC_PROG_CC(cc gcc)
 AC_PROG_CXX(c++ g++)
 AM_PROG_AS(as gas)
 
+AX_CHECK_COMPILE_FLAG([-std=c11], [AM_CFLAGS="$AM_CFLAGS -std=c11"])
+
 # Configure options.
 AM_OPTIONS
 
@@ -31,13 +33,6 @@ LT_PREREQ(2.2.6)
 LT_INIT([pic-only])
 AC_SUBST([LIBTOOL_DEPS])
 
-AX_CHECK_COMPILE_FLAG([-std=c11], [
-  CFLAGS+=" -std=c11"
-], [
-  echo "C compiler cannot compile C11 code"
-  exit -1
-])
-
```
I'd be inclined to not `#ifdef WT_CACHE_LINE_PAD_START/END` (I didn't need the an #ifdef if I specified `std=c11`), and I'd move `WT_CACHE_LINE_PAD_START/END` from `misc.h` to `hardware.h` (with `WT_CACHE_LINE_ALIGNMENT`).
```
diff --git a/src/include/hardware.h b/src/include/hardware.h
index 008bc55..34c2c1b 100644
--- a/src/include/hardware.h
+++ b/src/include/hardware.h
@@ -55,3 +55,19 @@
 #else
 #define        WT_CACHE_LINE_ALIGNMENT 64
 #endif
+
+/*
+ * Pad a structure so an array of structures get separate cache lines.
+ *
+ * Note that we avoid compiler structure alignment because that requires
+ * allocating aligned blocks of memory, and alignment pollutes any other type
+ * that contains an aligned field.  It is possible that a hot field positioned
+ * before this one will be on the same cache line, but not if it is also
+ * padded.
+ *
+ * This alignment has a small impact on portability as well, as we are using an
+ * anonymous union here, which is supported under C11 and earlier versions of
+ * the GNU standard.
+ */
+#define        WT_CACHE_LINE_PAD_START union { struct {
+#define        WT_CACHE_LINE_PAD_END   }; char __padding[WT_CACHE_LINE_ALIGNMENT]; };
diff --git a/src/include/misc.h b/src/include/misc.h
index c6cd168..66d4349 100644
--- a/src/include/misc.h
+++ b/src/include/misc.h
@@ -122,26 +122,6 @@
 #endif
 
 /*
- * Pad a structure so an array of structures get separate cache lines.
- *
- * Note that we avoid compiler structure alignment because that requires
- * allocating aligned blocks of memory, and alignment pollutes any other type
- * that contains an aligned field.  It is possible that a hot field positioned
- * before this one will be on the same cache line, but not if it is also
- * padded.
- *
- * This alignment has a small impact on portability as well, as we are using an
- * anonymous union here, which is supported under C11 and earlier versions of
- * the GNU standard.
- */
-#if __STDC_VERSION__ >= 201112L
-#define        WT_CACHE_LINE_PAD_START union { struct {
-#define        WT_CACHE_LINE_PAD_END   }; char __padding[WT_CACHE_LINE_ALIGNMENT]; };
-#else
-#define        WT_CACHE_LINE_PAD_START
-#define        WT_CACHE_LINE_PAD_END
-#endif
-/*
  * Flag set, clear and test.
  *
  * They come in 3 flavors: F_XXX (handles a field named "flags" in the structure
```

I'd rename `WT_CACHE_LINE_PAD_START/END` to `BEGIN/END`.

With that approach, gcc49, clang3.4, clang3.8, gcc5 and gcc6 all work for me. The build that failed was gcc47, gcc47 won't build on my system if `std=c11` is specified. I think we're probably OK with dropping support for gcc47?

Question, does this work if the structure we're aligning is larger than `WT_CACHE_LINE_ALIGNMENT`? Is that why we're no longer aligning `WT_SESSION_IMPL` structures? The original complaint in ticket WT-2402 was a misaligned access in a session structure, which makes me question if the change is sufficient.

And, I'm not sure why we removed the structure alignment from `test/format`?

**Comment by michaelcahill** [comment]
Thanks @daveh86 and @keithbostic for picking this up.

To the question about structs larger than a cache line: you are correct that this change will not pad them.  I don't see any advantage in padding {{WT_SESSION_IMPL}} because there's no chance of false sharing.  As far as the original complaint in the ticket, see the [Jenkins build](http://build.wiredtiger.com:8080/job/wiredtiger-sanitize-undefined/214/) referenced above.

**Comment by keithbostic** [comment]
@daveh86, I've confirmed that nobody cares if we drop support for gcc47, I think we're good to go with this approach.

I would suggest we put back in some kind of assertion about the structures we think are padded, I think there's some risk, especially with older compilers that `std=c11` isn't doing what we expect. Could we add something like:
```
WT_STATIC_ASSERT(sizeof(WT_SESSION_IMPL) % WT_CACHE_LINE_ALIGNMENT == 0);
```
to `build_verify.h`?

**Comment by michaelcahill** [comment]
@keithbostic why do we care if `sizeof(WT_SESSION_IMPL)` is a multiple of the cache line size?  As long as it is larger than a cache line, we can't have false sharing of hot fields...

**Comment by keithbostic** [comment]
@michaelcahill, sorry, didn't mean to use `WT_SESSION_IMPL` in that example, I just meant it as an example of the assert to add to `build_verify.h` to ensure this change is actually working correctly.

**Comment by keithbostic** [comment]
@daveh86, I pushed a few more changes, and I'm good with this branch once you're good with it, feel free to merge at will.

**Comment by daveh86** [comment]
Thanks @keithbostic, those few extra changes LGTM.

File: src/include/msvc.h [hunk 1]
```diff
@@ -19,8 +19,6 @@
 /*
  * Add MSVC-specific attributes and pragmas to types and function declarations.
  */
-#define	WT_COMPILER_TYPE_ALIGN(x)	__declspec(align(x))
```
Comment (agorrod) [inline src/include/msvc.h]: Have you tested that the new scheme works on Windows? I guess it should, but I'd rather find out before merging.

Comment (daveh86) [inline src/include/msvc.h]: The pull request tester with the anonymous unions has run on windows and passed. Anonymous unions are supported by Windows as far back as MSVC from 2003. Is there anything specific you were looking for in testing?

Comment (daveh86) [inline src/include/msvc.h]: The code in ```verify_build.h``` isn't run at all, as far as I can see.

Comment (daveh86) [inline src/include/msvc.h]: A little more digging and that function is never call, it simply has a bunch of static asserts in it. It is included as part of wt_internal.h, so should be compiled on Windows.

**Pull request merged**